docs/spec/auth/oauth.md

```diff
@@ -188,3 +194,160 @@ Content-Type: application/json
+                   |      +-------->      |    GET /v2/ (1)           |Registry|
+                   |      |        |      +--------------------------->        |
+                   |      |        |      |    401 Challenge          |        |
+                   |      |        |      <---------------------------+        |
+                   |      |        |      |                           |        |
+                   |      |        |      |                 +------+  |        |
+                   |      |        |      |  HEAD /token    |Token |  |        |
+                   |      |        |      +----------------->Server|  |        |
++-------+          |      | OAuth2 |      |  401 Challenge  |      |  |        |
+|  Web  | Redirect |      | Config |      <-----------------+      |  |        |
+|Browser| Web login|      <--------+      |                 |      |  |        |
+|       <----------+      | Login  |      |                 |      |  |        |
+|       | Auth Code|      | With   |      |                 |      |  |        |
+|       +---------->      | Code   |      |                 |      |  |        |
+|       |          |      +-------->      |  POST /token    |      |  |        |
+|       |          |      |        |      +----------------->      |  |        |
++-------+          |      |        |      |  Auth Tokens (2)|      |  |        |
+                   |      |        |      <-----------------+      |  |        |
+                   |      |        |      |                 |      |  |        |
+                   |      |        |      |                 +------+  |        |
+                   |      |        |      |  GET /v2/ (3)             |        |
+                   |      |        |      +--------------------------->        |
+                   |      | Login  |      |  OK                       |        |
+                   |      | Success|      <---------------------------+        |
+                   |      <--------+      |                           |        |
+                   |      |        |      |                           +--------+
+                   +------+        +------+
+
+(1) No authorization headers should be included
+(2) May return refresh token if offline access requested, always access token
+(3) Access token from previous request used to validate login
+```
+
+### OAuth2 Flow With Credentials
+
+The Docker client first prompts the user for credentials before
+calling the login endpoint on the daemon. The credentials are
+sent on login and when an oauth2 challenge is received from the
+token server, the credentials are immediately used with the
+`password` grant type.
+
+```
++------+        +------+
+|Docker| Login  |Docker|
+|Client| With   |Daemon|                           +--------+
+|      | Creds  |      |                           |Registry|
+|      +-------->      |    GET /v2/               |        |
```

Comment:
Let's make it clear that this should work on _any_ endpoint and not just `/v2/`.